### PR TITLE
Pre-execution histograms

### PR DESCRIPTION
**Pull request by Liorgal28**
On this MR I have added histograms in all main methods on the pre-execution flow in order to measure timings inside each method.
We can use these measurements to learn where the bottlenecks might be and if there are any.

### CODE REVIEW

File: bftengine/src/preprocessor/PreProcessor.cpp [hunk 1]
```diff
@@ -241,6 +241,7 @@ void PreProcessor::sendRejectPreProcessReplyMsg(NodeIdType clientId,
 
 template <>
 void PreProcessor::onMessage<ClientPreProcessRequestMsg>(ClientPreProcessRequestMsg *msg) {
+  concord::diagnostics::TimeRecorder scoped_timer(*histograms_.onMessage);
```
Comment (yuliasherman) [inline bftengine/src/preprocessor/PreProcessor.cpp:244]: 'onMessage' is a template function, do we combine results for all kind of received messages?

File: bftengine/src/preprocessor/PreProcessor.cpp [hunk 2]
```diff
@@ -470,6 +471,7 @@ void PreProcessor::finalizePreProcessing(NodeIdType clientId) {
   {
     lock_guard<mutex> lock(clientEntry->mutex);
     auto &reqProcessingStatePtr = clientEntry->reqProcessingStatePtr;
+    concord::diagnostics::TimeRecorder scoped_timer(*histograms_.finalizePreProcessing);
```
Comment (yuliasherman) [inline bftengine/src/preprocessor/PreProcessor.cpp:474]: This needs to be added at the beginning of the function to include wait-for-a-lock time, too.

File: bftengine/src/preprocessor/PreProcessor.hpp [hunk 3]
```diff
@@ -175,6 +176,39 @@ class PreProcessor {
   concordUtil::Timers::Handle requestsStatusCheckTimer_;
   const uint64_t preExecReqStatusCheckPeriodMilli_;
   concordUtil::Timers &timers_;
+
+  // 5 Minutes
+  static constexpr int64_t MAX_VALUE_MICROSECONDS = 1000 * 1000 * 60 * 5l;
+  // 60 seconds
+  static constexpr int64_t MAX_VALUE_NANOSECONDS = 1000 * 1000 * 1000 * 60l;
```
Comment (yuliasherman) [inline bftengine/src/preprocessor/PreProcessor.hpp:183]: Remove it

File: bftengine/src/preprocessor/PreProcessor.hpp [hunk 4]
```diff
@@ -175,6 +176,39 @@ class PreProcessor {
   concordUtil::Timers::Handle requestsStatusCheckTimer_;
   const uint64_t preExecReqStatusCheckPeriodMilli_;
   concordUtil::Timers &timers_;
+
+  // 5 Minutes
+  static constexpr int64_t MAX_VALUE_MICROSECONDS = 1000 * 1000 * 60 * 5l;
```
Comment (yuliasherman) [inline bftengine/src/preprocessor/PreProcessor.hpp:181]: Put just a number. 
Add a comment with a value in seconds.

File: bftengine/src/preprocessor/PreProcessor.cpp (hunk 1 above)
Comment (Liorgal28) [reply bftengine/src/preprocessor/PreProcessor.cpp:244]: If I understood you correctly, it is only for ClientPreProcessRequestMsg.

File: bftengine/src/preprocessor/PreProcessor.cpp (hunk 2 above)
Comment (Liorgal28) [reply bftengine/src/preprocessor/PreProcessor.cpp]: Done.

File: bftengine/src/preprocessor/PreProcessor.hpp (hunk 3 above)
Comment (Liorgal28) [reply bftengine/src/preprocessor/PreProcessor.hpp]: Done.

File: bftengine/src/preprocessor/PreProcessor.hpp (hunk 4 above)
Comment (Liorgal28) [reply bftengine/src/preprocessor/PreProcessor.hpp]: Done.

File: bftengine/src/preprocessor/PreProcessor.hpp [hunk 5]
```diff
@@ -177,21 +177,24 @@ class PreProcessor {
...
+      try {
+        registrar.perf.registerComponent("pre-execution",
+                                         {{"onMessage", onMessage},
+                                          {"launchReqPreProcessing", launchReqPreProcessing},
+                                          {"handlePreProcessedReqByNonPrimary", handlePreProcessedReqByNonPrimary},
+                                          {"handlePreProcessedReqByPrimary", handlePreProcessedReqByPrimary},
+                                          {"sendPreProcessRequestToAllReplicas", sendPreProcessRequestToAllReplicas},
+                                          {"finalizePreProcessing", finalizePreProcessing}});
+      } catch (std::invalid_argument &e) {
+        // if component already exist lets keep record on the same histograms
```
Comment (yuliasherman) [inline bftengine/src/preprocessor/PreProcessor.hpp:196]: Nit: exist-> exists

**Review by yuliasherman — APPROVED** [review]
LGTM

**Review by andrewjstone — APPROVED** [review]
Looks great! Thanks @Liorgal28

File: bftengine/src/preprocessor/PreProcessor.hpp [hunk 6]
```diff
@@ -175,6 +176,42 @@ class PreProcessor {
...
+      try {
+        registrar.perf.registerComponent("pre-execution",
+                                         {{"onMessage", onMessage},
+                                          {"launchReqPreProcessing", launchReqPreProcessing},
+                                          {"handlePreProcessedReqByNonPrimary", handlePreProcessedReqByNonPrimary},
+                                          {"handlePreProcessedReqByPrimary", handlePreProcessedReqByPrimary},
+                                          {"sendPreProcessRequestToAllReplicas", sendPreProcessRequestToAllReplicas},
+                                          {"finalizePreProcessing", finalizePreProcessing}});
+      } catch (std::invalid_argument &e) {
+        // if component already exists lets keep record on the same histograms
```
Comment (andrewjstone) [inline bftengine/src/preprocessor/PreProcessor.hpp:196]: I didn't notice this in my initial review. This doesn't work as intended. Instead, measurements will be silently dropped as I noted when updating the API in [this commit](https://github.com/vmware/concord-bft/pull/949/commits/b38c5fd850df55b70db95b46478947593488eb81) in [the tls pr](https://github.com/vmware/concord-bft/pull/949).

Unfortunately this is a bug that causes tests to fail from dual registration. I'm attempting to fix it, but the change might require input from you @Liorgal28 as it's a semantic issue. 

In short, there can only be one component with a given name. Ignoring the exception is no longer going to be an option because the process will terminate.

Comment (andrewjstone) [reply bftengine/src/preprocessor/PreProcessor.hpp:196]: I actually was able to fix this by modifying the test to clear the registrations. This is in general a problem of singletons, in that they introduce these types of side effects. Maybe in the future when we have a better startup sequence we can pass the registrar to components rather than using a singleton. For now this works fine and is easy enough.

Comment (Liorgal28) [reply bftengine/src/preprocessor/PreProcessor.hpp:196]: Thanks for addressing that issue, I commented in https://github.com/vmware/concord-bft/pull/949.
